File: 1743.restore-the-array-from-adjacent-pairs.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>

class Solution
{
public:
    std::vector<int> restoreArray(std::vector<std::vector<int>>& adjacentPairs)
    {
        // Collect the neighbours for lookups
        std::unordered_map<int, std::vector<int>> neighbours;
        for (const auto &p : adjacentPairs)
        {
            neighbours[p[0]].push_back(p[1]);
            neighbours[p[1]].push_back(p[0]);
        }

        // Find an end: just one neighbour
        auto it = std::find_if(neighbours.begin(), neighbours.end(),
                [](auto &n) { return n.second.size() == 1; });

        // Collect the chain by going through the neighbours
        std::vector<int> res;
        res.reserve(adjacentPairs.size() + 1);
        do
        {
            int num = it->first;
            int next = it->second[0];
            if (res.size() > 0 && next == res.back())
                next = it->second[1];
            res.push_back(num);
            it = neighbours.find(next);
        }
        while (it->second.size() > 1);
        res.push_back(it->first);
        return res;
    }
};

File: 1743.restore-the-array-from-adjacent-pairs.rs
use std::collections::{HashMap, HashSet};

impl Solution {
    pub fn restore_array(adjacent_pairs: Vec<Vec<i32>>) -> Vec<i32> {
        // Collect the neighbours for lookups
        let mut neighbours: HashMap<i32, Vec<i32>> = HashMap::new();
        for p in &adjacent_pairs {
            neighbours.entry(p[0]).or_insert(vec![]).push(p[1]);
            neighbours.entry(p[1]).or_insert(vec![]).push(p[0]);
        }

        // Find an end: just one neighbor
        let (&start, _) = neighbours.iter().find(|(_, v)| v.len() == 1).unwrap();

        // Collect the chain by going through the neighbours
        let mut res = Vec::with_capacity(adjacent_pairs.len() + 1);
        let mut current = start;
        res.push(current);
        loop {
            let ways = &neighbours[&current];
            let next = ways[0];
            if res.len() > 1 && next == res[res.len() - 2] {
                if ways.len() == 1 {
                    break;
                }
                current = ways[1];
            } else {
                current = next;
            }
            res.push(current);
        }

        res
    }
}
